### PR TITLE
refactor: convert packages/streams to strict TypeScript

### PR DESCRIPTION
**Pull request by dirkwa**
## Summary

- Convert all 31 JavaScript files in `packages/streams` from `util.inherits()` prototype pattern to ES6 classes with strict TypeScript
- Replace shared `StreamsApp` interface with per-stream inline app types so each module explicitly documents what it needs
- Add 55-test unit test suite for the streams package (previously zero coverage)

## Details

The streams package was the last major JS-only package in the server. This converts every file to strict TypeScript (`strict: true`, `noUncheckedIndexedAccess: true`) with proper ES6 classes, typed interfaces, and no `any` types.

**Per-stream inline app types:** Instead of a shared `StreamsApp` interface, each stream module declares exactly which `app` properties it uses via an inline type on its Options interface. This makes the contract between each stream and the app object self-documenting.

**Infrastructure:**

- `packages/streams/tsconfig.json` with strict compiler settings
- `package.json` updated with `dist/` paths, subpath exports map, and build/test scripts
- Root `tsconfig.json` updated with project reference
- ESLint override for `no-require-imports` (synchronous dynamic imports in constructors)
- Vendor declarations in `vendor.d.ts` for 9 untyped npm packages

**Consumer update:** `src/pipedproviders.ts` now uses `mod.default ?? mod` to handle both CJS and default exports from the dynamic `require()` calls.

**Test suite:** 55 tests across 14 test files covering:

- Pure transforms: liner, splitting-liner, replacer, from_json, keys-filter, timestamp-throttle
- Protocol converters: canboatjs, nmea0183-signalk, n2k-signalk (using real parsing libraries)
- Transport streams: execute, tcp, udp (using real I/O — TCP servers, UDP sockets, child processes)
- Logging and autodetect

## Tested

- `npx tsc --noEmit` — strict TypeScript compiles clean
- `npm run build:all` — all workspace packages compile
- `npm test` — 380 existing tests pass
- `npm run test -w @signalk/streams` — 55 new stream tests pass
- `eslint` + `prettier` — zero errors
- Live server test with N2K hardware

## Note

I rechecked the code several times but due to the nature of this massive change please allow beta test period before production.

### CODE REVIEW

**Comment by sbender9** [comment]
This throttling will only affect canboat, not canboatjs. It's very rare for anyone to use canboat.

**Comment by dirkwa** [comment]
> This throttling will only affect canboat, not canboatjs. It's very rare for anyone to use canboat.

Wow, that was quick. Seems you are sitting next to me :). 
Leaving out commit `82dc7bc` is "only" the TS refactor. 

I will look into canboat.js now.

**Comment by dirkwa** [comment]
> This throttling will only affect canboat, not canboatjs. It's very rare for anyone to use canboat.

Thanks for the quick catch! You're absolutely right — the rate limiting was in `execute.ts` which only covers the external canboat binary path.

I've moved it to `n2k-signalk.ts` where the `nmea2000JsonOut` event is actually emitted (the single source for all N2K output). This way it applies to all consumers — canboatjs-based devices (iKonvert, YDWG-02, W2K-01, Actisense, canbus) as well as the external binary.

See commit 2c45c0f7.

**Comment by sbender9** [comment]
This is not going to cover everything. The event can be emitted in other places. Sk-to-nmea2000, etc. 

I think this probably belongs in canboatjs.  

Also probably should not be a part of this PR.

**Comment by dirkwa** [comment]
> I think this probably belongs in canboatjs.
> 
> Also probably should not be a part of this PR.

Done as requested. TS conversion only. I leave issue #2305 with you. 
Thanks again and all the best!

File: packages/streams/src/n2k-signalk.ts [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  unknowCanName?: boolean
+  [key: string]: unknown
+}
+
+interface NotificationEntry {
+  lastTime: number
+  interval: ReturnType<typeof setInterval>
+}
+
+export default class ToSignalK extends Transform {
```
Comment (tkurki) [inline packages/streams/src/n2k-signalk.ts]: Naming seems odd, maybe n2kToSignalK?

File: packages/streams/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  enabled: boolean
+}
+
+export type CreateDebug = (namespace: string) => DebugLogger
+
+export interface DeltaCache {
+  setSourceDelta(key: string, delta: object): void
+}
+
+export interface StreamsApp extends EventEmitter {
```
Comment (tkurki) [inline packages/streams/src/types.ts]: There seems to be several properties here that are not even used and seem out of place:
- selfId
- streambundle
- savePluginOptions

I am not sure how useful having this "combined" streams application is though in the first place. Now each stream constructor needs an instance of StreamsApp....but they are all constructed from pipedprovider as any anyway, so I don't quite see the point. Maybe a better approach would be to have an inline type for each stream's `app` in the XXOptions type?

Comment (dirkwa) [reply packages/streams/src/types.ts]: You're right — `selfId`, `streambundle`, and `savePluginOptions` are not used by any stream. I'll remove them.

On the broader question of shared vs per-stream types: I agree that per-stream inline types would be more precise. The challenge is that `pipedproviders.ts` does dynamic `require()` dispatch — it constructs whichever stream class the config names at runtime, so the `app` it passes is inherently `unknown` at the TypeScript boundary regardless of how we type the stream side. The shared `StreamsApp` at least documents what the package expects from the app, even if the contract can't be enforced at the dynamic dispatch point.

Happy to go either direction. For now I'll clean up the unused properties. Tightening the `pipedproviders.ts` boundary would need a different architecture (static imports or a factory registry), which feels like a separate effort.

Comment (tkurki) [reply packages/streams/src/types.ts]: If each provider would have an inline type for the properties it expects it would document the contract it has with the `app` object, even if pipedprovider passes `app` as `any`. To me that would be more valuable than having the current "union" type that is not verified anywhere and it is unclear from the usage what properties each provider uses.

**Comment by tkurki** [comment]
Needs rebasing & changes after #2332

File: packages/streams/src/timestamp-throttle.test.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    throttle.write({ timestamp: String(now + 500) })
+
+    setTimeout(() => {
+      expect(results).to.have.length(1)
+    }, 100)
+
+    setTimeout(() => {
+      expect(results).to.have.length(2)
+      done()
+    }, 700)
```
Comment (tkurki) [inline packages/streams/src/timestamp-throttle.test.ts]: This adds almost a second to the test run, we could probably have stable result with a lot smaller timeouts.
